### PR TITLE
possible solution for global execution

### PR DESCRIPTION
**Pull request by samsonasik**
Ref https://github.com/rectorphp/rector/issues/6945

@MGatner could you verify if this works?

Note: its need to verify on e2e on rector/rector.

I am creating PR from mobile so probably missing space/typo.

### CODE REVIEW

**Comment by MGatner** [comment]
Yes this seems to solve it! I will test in a different environment as well.

**Comment by MGatner** [comment]
The autoload fix also works in PHP 8.1, though I am still getting the second error with `RemoveParentCallWithoutParentRector`: local version leaves the files alone while the global version removes the parent calls. Maybe this was two separate issues that both happened to have global/local differentiation?

Screen Shot 2022-01-17 at 11 10 56 AM

**Comment by samsonasik** [comment]
@MGatner it seems using direct `vendor/autoload.php` solve it https://github.com/rectorphp/rector-src/pull/1692/commits/465550308797b3716dc66f695d701fc6107db921

```diff
         $this->loadIfExistsAndNotLoadedYet(__DIR__ . '/../../../autoload.php');
-        if (is_dir(getcwd() . '/vendor/rector/rector')) {
-            $this->loadIfExistsAndNotLoadedYet(getcwd() . '/vendor/autoload.php');
-        }
+        $this->loadIfExistsAndNotLoadedYet('vendor/autoload.php');
```

Could you retry with above change https://github.com/rectorphp/rector-src/pull/1692/commits/465550308797b3716dc66f695d701fc6107db921 ?

**Comment by samsonasik** [comment]
I tried manually add to rector/rector scoped and it seems still cause error:

```bash
➜  parse-php7-code git:(main) ✗ ../../bin/rector --clear-cache
 0/1 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%PHP Fatal error:  Array and string offset access syntax with curly braces is no longer supported in /Users/samsonasik/www/rector/e2e/parse-php7-code/src/Foo.php on line 10
Fatal error: Array and string offset access syntax with curly braces is no longer supported in /Users/samsonasik/www/rector/e2e/parse-php7-code/src/Foo.php on line 10
```

so the autoload seems still overlapped.

**Comment by samsonasik** [comment]
Try with `is_dir()` first without `getcwd()` or `__DIR__` constant seems solve the e2e https://github.com/rectorphp/rector-src/pull/1692/commits/4608b9a5c73d7b0e7713b2f8d58558a0806cbcf8

Please verify latest change https://github.com/rectorphp/rector-src/pull/1692/commits/4608b9a5c73d7b0e7713b2f8d58558a0806cbcf8 again .

**Comment by samsonasik** [comment]
I retried in `codeigniter4-handlers` and it seems still apply `parent::tearDown()` removal, so it still not works:

```bash
➜  codeigniter4-handlers git:(develop) ✗ rector process --dry-run --clear-cache
 21/21 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%
2 files with changes
====================

1) tests/CommandsTest.php:29

    ---------- begin diff ----------
@@ @@

     protected function tearDown(): void
     {
-        parent::tearDown();
-
         stream_filter_remove($this->streamFilter);
     }
    ----------- end diff -----------

Applied rules:
 * RemoveParentCallWithoutParentRector


2) tests/CacheTest.php:9

    ---------- begin diff ----------
@@ @@
 {
     protected function tearDown(): void
     {
-        parent::tearDown();
-
         $this->factory->clearCache();
     }
    ----------- end diff -----------

Applied rules:
 * RemoveParentCallWithoutParentRector


                                                                                                                        
 [OK] 2 files would have changed (dry-run) by Rector                                                                    
                                                                                                                        
```

File: bin/rector.php [hunk 1]
```diff
@@ -117,15 +117,17 @@ public function autoloadFromCommandLine(): void
 
     public function loadIfExistsAndNotLoadedYet(string $filePath): void
     {
-        if (! file_exists($filePath)) {
+        $filePath = realpath($filePath);
```
Comment (MGatner) [inline bin/rector.php]: I don't know if Rector ever uses VFS for testing, but `realpath()` always fails on virtual paths.

Comment (samsonasik) [reply bin/rector.php]: I will rollback it as it cause error back on e2e test which look up from parent `../../bin` , probably need check if file path not contains `..`

Comment (samsonasik) [reply bin/rector.php]: @MGatner rolled back and check parent https://github.com/rectorphp/rector-src/pull/1692/commits/619e987538f29d145c861f540047b90e6f781720

Please verify again, I tried verify manually at my local dev and it seems working ok

File: bin/rector.php [hunk 2]
```diff
@@ -94,7 +94,12 @@ public function includeDependencyOrRepositoryVendorAutoloadIfExists(): void
      */
     public function autoloadProjectAutoloaderFile(): void
     {
-        $this->loadIfExistsAndNotLoadedYet(__DIR__ . '/../../../autoload.php');
+        if (is_dir(__DIR__ . '/../../vendor/rector/rector')) {
+            $this->loadIfExistsAndNotLoadedYet(__DIR__ . '/../../../autoload.php');
+            return;
+        }
+
+        $this->loadIfExistsAndNotLoadedYet('vendor/autoload.php');
```
Comment (samsonasik) [inline bin/rector.php]: it seems back make e2e looking parent `../../` bin error:

```bash
../../bin/rector --clear-cache
 0/1 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%PHP Fatal error:  Array and string offset access syntax with curly braces is no longer supported in /Users/samsonasik/www/rector/e2e/parse-php7-code/src/Foo.php on line 10
Fatal error: Array and string offset access syntax with curly braces is no longer supported in /Users/samsonasik/www/rector/e2e/parse-php7-code/src/Foo.php on line 10
```

**Comment by samsonasik** [comment]
@MGatner please verify again latest change, I tried locally and it seems working with add:

```php
        if (dirname(__DIR__) !== dirname(getcwd(), 2) && ! is_dir('vendor/rector/rector')) {
            $this->loadIfExistsAndNotLoadedYet('vendor/autoload.php');
        }
```

**Comment by samsonasik** [comment]
I tested with docker on getrector.org source and it seems working ok

**Comment by samsonasik** [comment]
@MGatner let's give it a try

**Pull request merged**